File: web/native-webrtc/ios/Sources/NativeWebRTCPlugin/FirestoreClient.swift
import Foundation

// Minimal Firestore REST client, authenticated with a Firebase ID token (not
// the native Firebase SDK — see NativeWebRTCPlugin.swift for why). Only
// implements the handful of operations the CallKit-answer path needs:
// read a document, partial-update fields, and add/list subcollection docs.
enum FirestoreClient {
    static let projectId = "chatapp-48786"
    private static var baseURL: String {
        "https://firestore.googleapis.com/v1/projects/\(projectId)/databases/(default)/documents"
    }

    enum ClientError: Error { case noToken, badResponse, http(Int, String) }

    private static func authToken() throws -> String {
        guard let token = UserDefaults.standard.string(forKey: "firebase_id_token"), !token.isEmpty else {
            throw ClientError.noToken
        }
        return token
    }

    private static func request(_ url: URL, method: String, body: Data? = nil) async throws -> Data {
        var req = URLRequest(url: url)
        req.httpMethod = method
        req.setValue("Bearer \(try authToken())", forHTTPHeaderField: "Authorization")
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = body
        let (data, response) = try await URLSession.shared.data(for: req)
        guard let http = response as? HTTPURLResponse else { throw ClientError.badResponse }
        guard (200...299).contains(http.statusCode) else {
            throw ClientError.http(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }
        return data
    }

    /// Fetches a document as a plain [String: Any] (Firestore's typed-value
    /// wrappers already unwrapped) or nil if it doesn't exist.
    static func getDocument(path: String) async throws -> [String: Any]? {
        let url = URL(string: "\(baseURL)/\(path)")!
        do {
            let data = try await request(url, method: "GET")
            guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let fields = json["fields"] as? [String: Any] else { return nil }
            return decodeFields(fields)
        } catch ClientError.http(404, _) {
            return nil
        }
    }

    /// Partial update (only the given top-level fields are touched — everything
    /// else on the document is left alone, matching Firestore's `updateDoc`).
    static func updateDocument(path: String, fields: [String: Any]) async throws {
        var urlStr = "\(baseURL)/\(path)?"
        urlStr += fields.keys.map { "updateMask.fieldPaths=\($0)" }.joined(separator: "&")
        let url = URL(string: urlStr)!
        let body = try JSONSerialization.data(withJSONObject: ["fields": encodeFields(fields)])
        _ = try await request(url, method: "PATCH", body: body)
    }

    /// Adds a new document to a subcollection with an auto-generated ID
    /// (mirrors the JS side's `addDoc`).
    static func addDocument(collectionPath: String, fields: [String: Any]) async throws {
        let url = URL(string: "\(baseURL)/\(collectionPath)")!
        let body = try JSONSerialization.data(withJSONObject: ["fields": encodeFields(fields)])
        _ = try await request(url, method: "POST", body: body)
    }

    /// Lists all documents in a subcollection, decoded the same way as getDocument.
    static func listDocuments(collectionPath: String) async throws -> [[String: Any]] {
        let url = URL(string: "\(baseURL)/\(collectionPath)")!
        let data = try await request(url, method: "GET")
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any],
              let docs = json["documents"] as? [[String: Any]] else { return [] }
        return docs.compactMap { doc in
            guard let fields = doc["fields"] as? [String: Any] else { return nil }
            var decoded = decodeFields(fields)
            decoded["__name"] = doc["name"] as? String
            return decoded
        }
    }

    // MARK: - Firestore typed-value <-> plain value

    private static func encodeValue(_ value: Any) -> [String: Any] {
        switch value {
        case let v as String: return ["stringValue": v]
        case let v as Bool: return ["booleanValue": v]
        case let v as Int: return ["integerValue": String(v)]
        case let v as Double: return ["doubleValue": v]
        case let v as [String: Any]: return ["mapValue": ["fields": encodeFields(v)]]
        case Optional<Any>.none: return ["nullValue": NSNull()]
        default: return ["stringValue": "\(value)"]
        }
    }

    private static func encodeFields(_ fields: [String: Any]) -> [String: Any] {
        var out: [String: Any] = [:]
        for (k, v) in fields { out[k] = encodeValue(v) }
        return out
    }

    private static func decodeValue(_ wrapped: Any) -> Any? {
        guard let dict = wrapped as? [String: Any] else { return nil }
        if let v = dict["stringValue"] as? String { return v }
        if let v = dict["booleanValue"] as? Bool { return v }
        if let v = dict["integerValue"] as? String { return Int(v) }
        if let v = dict["doubleValue"] as? Double { return v }
        if let map = dict["mapValue"] as? [String: Any], let f = map["fields"] as? [String: Any] {
            return decodeFields(f)
        }
        return nil
    }

    private static func decodeFields(_ fields: [String: Any]) -> [String: Any] {
        var out: [String: Any] = [:]
        for (k, v) in fields { out[k] = decodeValue(v) }
        return out
    }
}

File: web/android/app/src/main/java/app/chatapp/p2p/FirestoreClient.kt
package app.chatapp.p2p

import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Minimal Firestore REST client, authenticated with a Firebase ID token (not
 * the native Firebase SDK — mirrors web/native-webrtc's iOS FirestoreClient.swift
 * for the same reason: the native call engine needs to read/write the `calls/{id}`
 * doc without depending on the WebView/JS layer being alive). Only implements
 * what CallForegroundService needs: read a document, partial-update fields.
 */
object FirestoreClient {
    private const val PROJECT_ID = "chatapp-48786"
    private const val BASE_URL = "https://firestore.googleapis.com/v1/projects/$PROJECT_ID/databases/(default)/documents"

    class HttpError(val code: Int, val body: String) : Exception("HTTP $code: $body")

    private fun request(urlString: String, method: String, idToken: String, body: String? = null): String {
        val conn = URL(urlString).openConnection() as HttpURLConnection
        conn.requestMethod = method
        conn.setRequestProperty("Authorization", "Bearer $idToken")
        conn.setRequestProperty("Content-Type", "application/json")
        conn.connectTimeout = 10_000
        conn.readTimeout = 10_000
        if (body != null) {
            conn.doOutput = true
            conn.outputStream.use { it.write(body.toByteArray()) }
        }
        val code = conn.responseCode
        val stream = if (code in 200..299) conn.inputStream else conn.errorStream
        val text = stream?.bufferedReader()?.use { it.readText() } ?: ""
        if (code !in 200..299) throw HttpError(code, text)
        return text
    }

    /** Fetches a document as a plain JSONObject of unwrapped values, or null if 404. */
    fun getDocument(path: String, idToken: String): JSONObject? {
        return try {
            val json = JSONObject(request("$BASE_URL/$path", "GET", idToken))
            val fields = json.optJSONObject("fields") ?: return null
            decodeFields(fields)
        } catch (e: HttpError) {
            if (e.code == 404) null else throw e
        }
    }

    /** Partial update — only the given top-level fields are touched. */
    fun updateDocument(path: String, fields: Map<String, Any?>, idToken: String) {
        val mask = fields.keys.joinToString("&") { "updateMask.fieldPaths=$it" }
        val body = JSONObject().put("fields", encodeFields(fields)).toString()
        request("$BASE_URL/$path?$mask", "PATCH", idToken, body)
    }

    // MARK: - Firestore typed-value <-> plain value

    private fun encodeValue(value: Any?): JSONObject = when (value) {
        is String -> JSONObject().put("stringValue", value)
        is Boolean -> JSONObject().put("booleanValue", value)
        is Int -> JSONObject().put("integerValue", value.toString())
        is Double -> JSONObject().put("doubleValue", value)
        null -> JSONObject().put("nullValue", JSONObject.NULL)
        else -> JSONObject().put("stringValue", value.toString())
    }

    private fun encodeFields(fields: Map<String, Any?>): JSONObject {
        val out = JSONObject()
        for ((k, v) in fields) out.put(k, encodeValue(v))
        return out
    }

    private fun decodeValue(wrapped: JSONObject): Any? = when {
        wrapped.has("stringValue") -> wrapped.getString("stringValue")
        wrapped.has("booleanValue") -> wrapped.getBoolean("booleanValue")
        wrapped.has("integerValue") -> wrapped.getString("integerValue").toIntOrNull()
        wrapped.has("doubleValue") -> wrapped.getDouble("doubleValue")
        wrapped.has("mapValue") -> decodeFields(wrapped.getJSONObject("mapValue").optJSONObject("fields") ?: JSONObject())
        else -> null
    }

    private fun decodeFields(fields: JSONObject): JSONObject {
        val out = JSONObject()
        for (key in fields.keys()) {
            out.put(key, decodeValue(fields.getJSONObject(key)))
        }
        return out
    }
}
